client/CPlayerInterface.cpp

```diff
@@ -2637,9 +2637,39 @@ void CPlayerInterface::doMoveHero(const CGHeroInstance* h, CGPath path)
 		ETerrainType newTerrain;
 		int sh = -1;
 
-		const TerrainTile * curTile = cb->getTile(CGHeroInstance::convertPosition(h->pos, false));
+		auto allowedToStopAt = [this,h]{ //make sure hero won't able to stop on blocked tile while flying
+			bool canStopAt = true;
+			const TerrainTile * curTile = cb->getTile(CGHeroInstance::convertPosition(h->pos, false));

```

Comment:
All "allowed to" logic should belong to Callback, since you're calling it here anyway. Player interface should never allow players to stop in mid-air even if they insist to do so. In fact this condition should be checked on server side and request to stop moving would be ignored.